WelsonJS.Toolkit/EsentInterop/Windows8JetApi.cs

```diff
@@ -0,0 +1,833 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="Windows8JetApi.cs" company="Microsoft Corporation">
+//     Copyright (c) Microsoft Corporation.
+// </copyright>
+//-----------------------------------------------------------------------
+
+namespace Microsoft.Isam.Esent.Interop.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Runtime.InteropServices;
+    using Microsoft.Isam.Esent.Interop.Vista;
+    using Microsoft.Isam.Esent.Interop.Windows8;
+
+    /// <summary>
+    /// Windows8 calls to the ESENT interop layer. These calls take the managed types (e.g. JET_SESID) and
+    /// return errors.
+    /// </summary>
+    internal sealed partial class JetApi : IJetApi
+    {
+        #region Transactions
+        /// <summary>
+        /// Causes a session to enter a transaction or create a new save point in an existing
+        /// transaction.
+        /// </summary>
+        /// <param name="sesid">The session to begin the transaction for.</param>
+        /// <param name="userTransactionId">An optional identifier supplied by the user for identifying the transaction.</param>
+        /// <param name="grbit">Transaction options.</param>
+        /// <returns>An error if the call fails.</returns>
+        public int JetBeginTransaction3(JET_SESID sesid, long userTransactionId, BeginTransactionGrbit grbit)
+        {
+            TraceFunctionCall();
+            return Err(NativeMethods.JetBeginTransaction3(sesid.Value, userTransactionId, unchecked((uint)grbit)));
+        }
+
+        /// <summary>
+        /// Commits the changes made to the state of the database during the current save point
+        /// and migrates them to the previous save point. If the outermost save point is committed
+        /// then the changes made during that save point will be committed to the state of the
+        /// database and the session will exit the transaction.
+        /// </summary>
+        /// <param name="sesid">The session to commit the transaction for.</param>
+        /// <param name="grbit">Commit options.</param>
+        /// <param name="durableCommit">Duration to commit lazy transaction.</param>
+        /// <param name="commitId">Commit-id associated with this commit record.</param>
+        /// <returns>An error if the call fails.</returns>
+        public int JetCommitTransaction2(JET_SESID sesid, CommitTransactionGrbit grbit, TimeSpan durableCommit, out JET_COMMIT_ID commitId)
+        {
+            TraceFunctionCall();
+            this.CheckSupportsWindows8Features("JetCommitTransaction2");
+            int err;
+            uint cmsecDurableCommit = (uint)durableCommit.TotalMilliseconds;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

An integer overflow occurs when an arithmetic operation is performed on integer data types without proper checking of maximum value stored in that data type. As C# does not automatically check for or prevent overflow by default, rather it wraps around the result, which may lead to unexpected behaviour or incorrect results. Therefore perform input validation on any numeric input by ensuring that it is within the expected range, when signed integers are required, ensure that the range check includes maximum values. For more information on Integer Overflow refer to https://cwe.mitre.org/data/definitions/190.html.